File: onecard-server/src/modules/game/game.service.ts
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import type {
  GameSettings,
  GameState,
} from '@/modules/game/domain/types/gameState';
import { EngineStepResult } from '@/modules/game/domain/engine/gameEngine';
import {
  GameSessionRecord,
  GameStateStore,
} from '@/modules/game/state/game-state.store';
import { GameEngineService } from '@/modules/game/services/game-engine.service';
import {
  GameActionDto,
  GameActionType,
} from '@/modules/game/dto/game-action.dto';
import { GameAiService } from '@/modules/game/services/game-ai.service';

export interface GameResource {
  id: string;
  state: GameState;
  createdAt: string;
  updatedAt: string;
}

export interface GameSummary {
  id: string;
  createdAt: string;
  updatedAt: string;
  gameStatus: GameState['gameStatus'];
  currentPlayerIndex: number;
}

type StepResult = EngineStepResult;

@Injectable()
export class GameService {
  public constructor(
    private readonly gameStateStore: GameStateStore,
    private readonly gameEngine: GameEngineService,
    private readonly gameAiService: GameAiService,
  ) {}

  public listGames(): GameSummary[] {
    return this.gameStateStore.list().map((record) => ({
      id: record.id,
      createdAt: record.createdAt.toISOString(),
      updatedAt: record.updatedAt.toISOString(),
      gameStatus: record.state.gameStatus,
      currentPlayerIndex: record.state.currentPlayerIndex,
    }));
  }

  public createGame(settings?: Partial<GameSettings>): GameResource {
    const record = this.gameStateStore.create(settings);
    return this.toResource(record);
  }

  public getGame(gameId: string): GameResource {
    const record = this.findGameOrThrow(gameId);
    return this.toResource(record);
  }

  public applyAction(
    gameId: string,
    actionPayload?: GameActionDto,
  ): StepResult {
    if (!actionPayload) {
      throw new BadRequestException('Action payload is required');
    }

    const record = this.findGameOrThrow(gameId);
    if (
      record.state.gameStatus === 'waiting' &&
      actionPayload.type !== GameActionType.START_GAME
    ) {
      throw new BadRequestException('게임이 아직 시작되지 않았습니다.');
    }

    const action = this.gameEngine.buildAction(actionPayload);
    const result = this.gameEngine.step(record.state, action);
    this.gameStateStore.updateState(gameId, result.state);
    return result;
  }

  public executeAiTurn(gameId: string): StepResult {
    const record = this.findGameOrThrow(gameId);
    const currentState: GameState = record.state;
    if (currentState.gameStatus !== 'playing') {
      throw new BadRequestException('게임이 아직 시작되지 않았습니다.');
    }
    if (!this.gameAiService.isAiTurn(currentState)) {
      throw new BadRequestException('현재 차례는 AI가 아닙니다.');
    }

    const aiResult: StepResult | null =
      this.gameAiService.playWhileAiTurn(currentState);
    if (!aiResult) {
      throw new BadRequestException('AI가 수행할 수 있는 행동이 없습니다.');
    }

    this.gameStateStore.updateState(gameId, aiResult.state);
    return aiResult;
  }

  public deleteGame(gameId: string): void {
    const deleted = this.gameStateStore.delete(gameId);
    if (!deleted) {
      throw new NotFoundException(`Game ${gameId} not found`);
    }
  }

  private findGameOrThrow(gameId: string): GameSessionRecord {
    const record = this.gameStateStore.find(gameId);
    if (!record) {
      throw new NotFoundException(`Game ${gameId} not found`);
    }
    return record;
  }

  private toResource(record: GameSessionRecord): GameResource {
    return {
      id: record.id,
      state: record.state,
      createdAt: record.createdAt.toISOString(),
      updatedAt: record.updatedAt.toISOString(),
    };
  }
}

File: onecard-server/src/onecard_api/services/game_service.py
from __future__ import annotations

from fastapi import HTTPException, status

from onecard_api.domain.card_utils import is_valid_play
from onecard_api.domain.engine import GameAction
from onecard_api.domain.types import GameState
from onecard_api.services.game_ai_service import GameAiService
from onecard_api.services.game_engine_service import GameEngineService
from onecard_api.services.game_state_store import GameSessionRecord, GameStateStore


class GameService:
    def __init__(
        self,
        game_state_store: GameStateStore,
        game_engine: GameEngineService,
        game_ai_service: GameAiService,
    ) -> None:
        self._game_state_store = game_state_store
        self._game_engine = game_engine
        self._game_ai_service = game_ai_service

    def list_games(self) -> list[dict]:
        return [
            {
                "id": record["id"],
                "createdAt": record["created_at"].isoformat(),
                "updatedAt": record["updated_at"].isoformat(),
                "gameStatus": record["state"]["gameStatus"],
                "currentPlayerIndex": record["state"]["currentPlayerIndex"],
            }
            for record in self._game_state_store.list()
        ]

    def create_game(self, settings: dict | None = None) -> dict:
        record = self._game_state_store.create(settings)
        return self._to_resource(record)

    def get_game(self, game_id: str) -> dict:
        record = self._find_game_or_throw(game_id)
        return self._to_resource(record)

    def apply_action(self, game_id: str, action_payload: dict | None) -> dict:
        if action_payload is None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Action payload is required",
            )
        record = self._find_game_or_throw(game_id)

        if (
            record["state"]["gameStatus"] == "waiting"
            and action_payload.get("type") != "START_GAME"
        ):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="게임이 아직 시작되지 않았습니다.",
            )

        if action_payload.get("type") == "PLAY_CARD":
            self._assert_playable_card(record["state"], action_payload)

        action: GameAction = self._game_engine.build_action(action_payload)
        result = self._game_engine.step(record["state"], action)
        self._game_state_store.update_state(game_id, result["state"])
        return result

    async def execute_ai_turn(self, game_id: str) -> dict:
        record = self._find_game_or_throw(game_id)
        current_state: GameState = record["state"]
        if current_state["gameStatus"] != "playing":
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="게임이 아직 시작되지 않았습니다.",
            )
        if not self._game_ai_service.is_ai_turn(current_state):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="현재 차례는 AI가 아닙니다.",
            )

        ai_result = await self._game_ai_service.play_while_ai_turn(
            current_state, {"gameId": game_id}
        )
        if ai_result is None:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="AI가 수행할 수 있는 행동이 없습니다.",
            )
        self._game_state_store.update_state(game_id, ai_result["state"])
        return ai_result

    def delete_game(self, game_id: str) -> None:
        deleted = self._game_state_store.delete(game_id)
        if not deleted:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Game {game_id} not found",
            )

    def _find_game_or_throw(self, game_id: str) -> GameSessionRecord:
        record = self._game_state_store.find(game_id)
        if not record:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Game {game_id} not found",
            )
        return record

    def _to_resource(self, record: GameSessionRecord) -> dict:
        return {
            "id": record["id"],
            "state": record["state"],
            "createdAt": record["created_at"].isoformat(),
            "updatedAt": record["updated_at"].isoformat(),
        }

    def _assert_playable_card(self, state: GameState, payload: dict) -> None:
        player_index = payload.get("playerIndex", -1)
        card_index = payload.get("cardIndex", -1)

        if not isinstance(player_index, int) or player_index < 0 or player_index >= len(state["players"]):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="잘못된 playerIndex 입니다.",
            )

        hand = state["players"][player_index]["hand"]
        if not isinstance(card_index, int) or card_index < 0 or card_index >= len(hand):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="잘못된 cardIndex 입니다.",
            )

        if len(state["discardPile"]) == 0:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="discardPile 이 비어 있습니다. 게임 상태를 확인하세요.",
            )

        played_card = hand[card_index]
        top_card = state["discardPile"][0]
        damage = state["damage"]
        if not is_valid_play(played_card, top_card, damage):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="규칙에 맞지 않는 카드입니다.",
            )
